### ISSUE TITLE
Model checkpoint load ignores wrong shape

### ISSUE THREAD

**Issue by albertz** [tensorflow/tensorflow#52220]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOSX 11.6 + Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip binary
- TensorFlow version (use command below):  v2.6.0-rc2-32-g919f693420e 2.6.0
- Python version: 3.9.6 + 3.8
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: None
- GPU model and memory: None

**Describe the current behavior**

Saving a variable of shape (3,3,40,32) into a checkpoint, and then loading the variable with shape (3,3,1,32) from this checkpoint does not cause an error. Instead, it loads fine. The variable shape still claims to be (3,3,1,32), but when evaluated, one can see that it is indeed (3,3,40,32).

So there does not seem to be any validation of the shape during checkpoint loading.

**Describe the expected behavior**

In general, if a tensor or variable claims to be of some static shape, I think it should never be possible that its actual shape when evaluated is different.

**Standalone code to reproduce the issue**
```

import tensorflow as tf

print("TF:", tf.__version__)
tf.compat.v1.disable_eager_execution()

filename = "test-ckpt-diff-shape.model"


with tf.Graph().as_default() as graph:
    with tf.compat.v1.Session(graph=graph) as session:
        shape1 = (3,3,40,32)
        v = tf.compat.v1.get_variable(name="W", shape=shape1)
        print(v)
        saver = tf.compat.v1.train.Saver(var_list=[v])
        session.run(tf.compat.v1.global_variables_initializer())
        saver.save(sess=session, save_path=filename)


with tf.Graph().as_default() as graph:
    with tf.compat.v1.Session(graph=graph) as session:
        shape2 = (3,3,1,32)
        v = tf.compat.v1.get_variable(name="W", shape=shape2)
        print(v)
        saver = tf.compat.v1.train.Saver(var_list=[v])
        saver.restore(sess=session, save_path=filename)
        v_raw = session.run(v)
        print(v)
        print(v_raw.shape)
        assert v.shape.as_list() == list(shape2)
        assert v.shape.as_list() == list(v_raw.shape)
```

**Other info / logs**

I stumbled upon this problem because it causes a seemingly unrelated error for some following 2D convolution where this variable is used as a kernel:
```
2021-09-30 12:52:25.768174: W tensorflow/core/framework/op_kernel.cc:1767] OP_REQUIRES failed at conv_ops.cc:1
115 : Not found: No algorithm worked!
TensorFlow exception: 2 root error(s) found.
  (0) Not found: No algorithm worked!
         [[node conv0/convolution (defined at u/schmitt/src/returnn/returnn/tf/layers/basic.py:4061) ]]
         [[output/rec/while/Switch_14/_553]]
  (1) Not found: No algorithm worked!
         [[node conv0/convolution (defined at u/schmitt/src/returnn/returnn/tf/layers/basic.py:4061) ]]
0 successful operations.
```
This was reported here: https://github.com/rwth-i6/returnn/issues/703

There are a couple of these errors also reported here:
* https://github.com/tensorflow/tensorflow/issues/43174
* https://github.com/tensorflow/tensorflow/issues/45044
* https://github.com/tensorflow/tensorflow/issues/48117

I wonder if some of them have a similar source.

But anyway, this issue is not really about the convolution error, but about the ckeckpoint loading behavior, and the variable shape.

**Comment by mohantym** [comment]
Hi @Albert-Z-Guo ,Tried to replicate this in 2.6 ! Assertion Error in **second Assertion** might be  raising due to **mismatch in new session shape (3,3,40,32) and  old session shape (3,3,1,32)** ,while  **first Assertion** new session shape is **being matched to declared shape (3,3,1,32)** .In Attaching [GIST](https://colab.research.google.com/gist/mohantym/1d70ba29c7ea96c84773e1e62fc0df9a/github_52220.ipynb) for reference.

**Comment by albertz** [comment]
Yes and this is wrong. You always must have `v.shape.as_list() == list(session.run(v).shape)`.

**Comment by albertz** [comment]
It's also wrong that the model checkpoint loading was done without error. There should have been an exception there.

**Comment by mohantym** [comment]
Hi @Saduf2019 , Could you please look into this issue . It is replicating in [2.5 ](https://colab.research.google.com/gist/mohantym/b68163371f9e8a869b0244b103ab0572/github_52220.ipynb#scrollTo=S8zOfkH3z28T),[2.6 ](https://colab.research.google.com/gist/mohantym/1d70ba29c7ea96c84773e1e62fc0df9a/github_52220.ipynb#scrollTo=YGsXUWna1zP_)and [nightly.](https://colab.research.google.com/gist/mohantym/fc67428fd6b301bc309c953991dee235/github_52220.ipynb#scrollTo=S8zOfkH3z28T)

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by albertz** [comment]
The [linked Gist from above](https://colab.research.google.com/gist/mohantym/1d70ba29c7ea96c84773e1e62fc0df9a/github_52220.ipynb) still shows the same bug with TF 2.17.0, so it is still relevant, still not fixed.
